numpy/core/src/umath/ufunc_object.c

```diff
@@ -2223,9 +2224,16 @@ _get_coredim_sizes(PyUFuncObject *ufunc, PyArrayObject **op,
     int nout = ufunc->nout;
     int nop = nin + nout;
 
-    for (j = 0; j < ufunc->core_num_dim_ix; ++j) {
-        /* support fixed-size dim names */
-        core_dim_sizes[j] = ufunc->core_dim_sizes[j];
+    if (ufunc->version > 0) {
+        for (j = 0; j < ufunc->core_num_dim_ix; ++j) {
+            /* support fixed-size dim names */
+            core_dim_sizes[j] = ufunc->core_dim_sizes[j];
+        }
+    }
+    else {
+        for (j = 0; j < ufunc->core_num_dim_ix; ++j) {
+            core_dim_sizes[j] = UFUNC_CORE_DIM_SIZE_UNSET;
```

Comment:
Instead of adding the loop for the version = 0 case, I would define the default in the definition:
```
 npy_intp inner_dimensions[NPY_MAXDIMS+1] = {UFUNC_CORE_DIM_SIZE_UNSET};
```
Although, actually I think this takes more time, since the initialization would be done on every call.

But all these loops threaten to make the code less readable. How about a `initialize_core_information(ufunc, core_num_dims, core_dim_sizes, core_dim_flags)` function that does the version checking inside?

Reply:
I am -1 on setting the values in the definition. Perhaps we could coalesce the two `if { for ... } else { for ...} into one, the indexing is the same. 

I am not sure an 8 line function improves readability, I always forget what the functions do, need to scroll around to find the definition, scan it to get a general idea what is up, scroll back to the use to put it in context, then go back to the definition of the function to actually understand what is done, but everyone has different tastes. Whatever you feel is more clear

Reply:
I think my main wish is for back-ward compatibility stuff to be hidden as much as possible.

Reply:
OK. Do you want me to redo this as a function in a PR here?